### ISSUE TITLE
Allow subclassing a sealed class outside the library with an abstract class

### ISSUE THREAD

**Issue by srawlins** [dart-lang/language#3135]
In the analyzer package we have a lot of sealed-in-spirit class hierarchies. The class declarations are found in two files, two libraries, one contains the public interfaces, and the other contains the private impls. Something like this:

```dart
// lib/foo.dart

abstract class Sealed {}
abstract class A implements Sealed {}
abstract class B implements Sealed {}

// lib/src/foo.dart <-- implementation private!

abstract class SealedImpl implements Sealed {}
class AImpl extends SealedImpl implements A {}
class BImpl extends SealedImpl implements B {}
```

(I hope you can trust that the hierarchy serves a purpose; i.e. SealedImpl has some implementation, etc.)

I think we would reap a lot of exhaustiveness benefit from sealing `Sealed`. But today we cannot change `Sealed` to be `sealed` because of that pesky `SealedImpl` in the private impl library. And we cannot move `SealedImpl` because (a) it has implementation, (b) it is not public API, and (c) that would remove any exhaustiveness benefits; it would be confusing to have to cover A, B, _and `SealedImpl`_ in any exhaustive switch.

This is frustrating: I can see that each concrete subclass of `Sealed` in fact _is a subtype_ of `A` or of `B`. The family of classes is sealed in practice. How to prove it to static analysis?

It occurs to me that I believe we can allow subclassing `sealed class Sealed` outside of its library by an abstract class. Today we have static checks that amount to the blanket rule:

* Given a sealed class _C1_ declared in library _L1_, and given a class _C2_ declared in library _L2_, which has _C1_ as a direct supertype, and given _L1 != L2_: we have a compile-time error.

I think we can alter that rule to be more nuanced:

* Given a sealed class _C1_ declared in library _L1_, and given a class _C2_ declared in library _L2_, which has _C1_ as a supertype (however indirectly), and given _L1 != L2_:
  * If _C2_ is abstract (either `sealed` or `abstract`), we do not have a compile-time error.
  * If _C2_ does not directly subclass any of the classes in the sealed family of _C1_, we have a compile-time error.

I think this maintains nicely the semantics of the sealed family (both in compile-time enforcement and conceptually):
* A sealed class has a (sealed) family of subclasses declared in its same library.
  * (Note for new logic: More abstract subclasses can be declared anywhere, but they don't contribute to the sealed family.)
* For every object whose runtime type, _T_, is a subtype of the sealed class, _T_ is also a subtype of (or equal to) some class in the sealed family.

In the `Sealed` hierarchy above, you can see maybe an exhaustiveness wrinkle: A developer might still use `SealedImpl` in a switch case and 🤷  maybe it is useful to do so, but such a case would... not contribute to exhaustiveness? `SealedImpl` is not part of the sealed family of `Sealed`, so maybe no wrinkle.

Mixins cannot be abstract so I think they're not relevant to this proposal.

**Comment by eernstg** [comment]
> we have a lot of sealed-in-spirit class hierarchies

I think it would make sense to enhance the terminology a bit in this area: Assume that a library _L_ defines a maximal set of classes `C1 .. Cn` that are subtypes of a given root `C`, and each of these classes have the modifier `sealed`, and every class on every superinterface path from `Cj` to `C` is one of the classes `C, C1 .. Cn`. In this case we say that `{C, C1 .. Cn}` is a _sealed hierarchy_. There may be any number of subtypes of each member of the sealed hierarchy in _L_, but the point is that every class in the sealed hierarchy is an always-exhaustive type (and we can find a minimal exhaustive set of subtypes for every type `D` in `C, C1 .. Cn` by finding all direct subtypes of `D` in _L_).

In the example [here](https://github.com/dart-lang/language/issues/3135#issue-1745011694), the intended sealed hierarchy is just the single class `Sealed`, as far as I can see. But I would assume that many subtype hierarchies with a sealed class `S` at the top would actually have a larger sealed hierarchy below `S`. The conceptual justification would be that every node in the sealed hierarchy plays the role as one member of a well-known set of "sibling" classes. In the example, class `Sealed` is known to consist of `A` and `B`; but, presumably `A` could be sealed and it could consist of `A1` and `A2` and `A3`, and similarly for `B`.

If the sealed hierarchy is indeed just one class then we should be able to create an `Impl` structure as follows:

```dart
// lib/foo.dart

sealed class Sealed {}
abstract class A implements Sealed {}
abstract class B implements Sealed {}

// lib/src/foo.dart <-- implementation private!

abstract class SealedImpl {} // Could be a mixin, as @rrousselGit mentioned.
class AImpl extends SealedImpl implements A {}
class BImpl extends SealedImpl implements B {}
```

`SealedImpl` can provide whatever is needed in terms of member implementations, and the fact that we got them all (in order to satisfy `implements Sealed`, which we can't actually write any more) is checked indirectly: Any missing implementations in `SealedImpl` will just translate into missing implementations in `AImpl` and/or `BImpl`.

In other words, there is no problem as long as the sealed hierarchy is a single class.

So let's assume that the sealed hierarchy is (at least) `Sealed`, `A`, and `B`.

```dart
// lib/foo.dart

sealed class Sealed {}
sealed class A implements Sealed {}
sealed class B implements Sealed {}
class A1 implements A {}
class A2 implements A {}
class B1 implements B {}
class B2 implements B {}

// lib/src/foo.dart <-- implementation private!

// Create a shadow of the sealed hierarchy. We don't get to declare subtype relationships,
// because that's an error, but we do provide implementations of various members.
abstract class SealedImpl /*would implement Sealed*/ {}
abstract class AImpl extends SealedImpl /*would implement A*/ {}
abstract class BImpl extends SealedImpl /*would implement B*/ {}

// Just below the sealed hierarchy we can have concrete classes,
// and at this point they can have the "correct" type.
class A1Impl extends AImpl implements A1 {}
class A2Impl extends AImpl implements A2 {}
class B1Impl extends BImpl implements B1 {}
class B2Impl extends BImpl implements B2 {}
```

It's inconvenient that we cannot declare the subtype relationships from `SealedImpl` to `Sealed`, from `AImpl` to `A`, and so on, but there is no real danger in that: The fact that `A1Impl` actually implements `A1` will provide all the needed subtype relationships, and it will give rise to compile-time errors if any members do not have an implementation, or they have a signature which is not a correct override of the member signature in the sealed hierarchy.

It may not be pretty, but I do think that it would work.

**Comment by srawlins** [comment]
Sorry, I believe it is useful (one might say imperative) to have `SealedImpl implements Sealed`. To flesh it out a bit more:

```dart
// lib/foo.dart

abstract class Sealed {
  Iterable<SyntacticEntity> get childEntities;
}

// lib/src/foo.dart <-- implementation private!

abstract class SealedImpl implements Sealed {
  @override
  Iterable<SyntacticEntity> get childEntities => _childEntities.syntacticEntities;
}
```

The `implements Sealed` provides all the usual guards, that members have the right types, the right names, etc.

WRT to my proposed subclassing rules, would they be technically difficult? Hinging on finding the sealed family when only given `abstract class C2 implements C1` and a sealed `C1`?

**Comment by lrhn** [comment]
> We would replace that rule by this one:
>
>> It's a compile-time error if [..] a concrete declaration _D_ from a library _L_ has a (direct or indirect) superdeclaration _S_ marked `sealed` [..] in a library different from _L_, and _D_ does not have a (direct or indirect) superdeclaration _U_ from the library _L_ such that _S_ is a direct superdeclaration of _U_.
>
> A _concrete declaration_ is a class or mixin class declaration which isn't abstract.

(or enum declaration.)

This basically boils down to:

* All *concrete* subclasses of a sealed class *S* must be subtypes of one of the immediate subclasses of *S* declared in the same library as *S*. 

That's a very direct specification of the requirement necessary for ensuring exhaustiveness, so it pretty certainly would work. Every other rule we've made is trying to enforce that restriction by disallowing some patterns (and preferably in a way where it's directly visible on a class declaration which constraints it imposes on its subclasses).

So when we say that you cannot subclass a `sealed` class from another library, not even with an `abstract` declaration, it's because would not be directly visible on that abstract subclass that there is a transitive constraint. You won't notice until you try to make a concrete subclass of the abstract class, possibly multiple classes down, at which point you're told "This concrete class must implement one of these classes: SealedSubclass1, SealedSubclass2, ..., SealedSubclassN, because it implements the sealed class Sealed". 

(And then, if any of those immediate subclasses are themselves sealed, we may want to find the non-sealed frontier below the sealed classes. And maybe omit `final` declarations and enums, since you can't implement those anyway. The class cannot be in the same library, because then it is one of the immediate subclasses, and it implements itself, so no problem.)

We could do that, it would be sound for exhaustiveness, but it causes a lower direct readability of the classes.

**Comment by scheglov** [comment]
FWIW, I *don't* want to make any API classes sealed.
This puts too much obligations on me.
So, no, I don't think that this is very valuable.

I find it very useful to use `sealed` classes for (so far) local, library private, classes.
https://dart-review.googlesource.com/c/sdk/+/306160/15/pkg/analysis_server/lib/src/services/refactoring/framework/write_invocation_arguments.dart#85
The theme: just because I can! :-P

So, I appreciate local type safety.
Maybe we could allow something better inside the package / module / etc.
But I'm not going to promise anything to my clients.

**Comment by srawlins** [comment]
Thanks a lot for all of the input! I like that it seems sound. I don't know of other cases that use our style of public-interface and private-impl (I mean, it's classic, I just don't think I've seen it outside our code base in a Dart package).

**Comment by rrousselGit** [comment]
> FWIW, I _don't_ want to make any API classes sealed. This puts too much obligations on me. So, no, I don't think that this is very valuable.

@scheglov You mean that you don't want `analyzer` to expose sealed classes in its interface?  
If so, why?

As an avid user of analyzer, it sounds very useful. I already wrote a bunch of analyzer `switch` statements which are begging to become exhaustive

**Comment by jakemac53** [comment]
> As an avid user of analyzer, it sounds very useful. I already wrote a bunch of analyzer `switch` statements which are begging to become exhaustive

Adding a new subtype is then breaking, which means even more breaking versions of analyzer. But you could view that as a plus as well - you will know exactly the places in your code that need to update and which cases need to be handled.

**Comment by scheglov** [comment]
Yes, I agree, that adding a method to `AstVisitor`, if we allowed implementing it to the clients, would have been a breaking change. In a sense visitors and sealed classes are similar. We don't allow implementing `AstVisitor`, just by saying this, and don't use any language syntax to enforce it. Instead, we say that clients should extend `SimpleAstVisitor` for example. This keeps the code compile-error free, when we add a method. But maybe not working as expected, if the analyzer adds a new node that replaces some other node, and now another method of the visitor is invoked. The client will not know about this.

So, there might be an argument for using sealed classes, and implementing `AstVisitor`.

But I'm curious about your specific cases that you noticed for exhaustive switches.

**Comment by rrousselGit** [comment]
By that standard, then pretty much everything is a breaking change.   
Like how adding a new type could break apps, because the app already has a class with that name, so now they have a name conflict. Or how adding a new property to a `final` class is still breaking because it could shadow extension methods.

Semantic versioning makes no sense with that considered.

**Comment by scheglov** [comment]
Sure, I understand `sealed` in general.
Do you mean that you would want `sealed` to tell you what to update?
I initially thought that you found cases where an exhaustive `switch` would be just more convenient, not only more correct.
Correctness is not a small thing, just if you have more arguments - bring them :-)

**Comment by rrousselGit** [comment]
> Sure, I understand `sealed` in general.
> Do you mean that you would want `sealed` to tell you what to update?
> I initially thought that you found cases where an exhaustive `switch` would be just more convenient, not only more correct.
> Correctness is not a small thing, just if you have more arguments - bring them :-)

I'll admit I don't understand your question and your usage of `convenient` vs `correct`. I'm not sure what they mean in this context. Happy to help by explaining more my point of view. But I don't quite understand your comment yet.


But anyway, for me, the Ast/Element tree is _the_ textbook definition of what sealed classes are for.  
While patiently waiting for sealed classes all these years, I've always been eagerly waiting for `analyzer` to use them. So this discussion breaks my heart a bit, ha

**Comment by lrhn** [comment]
>  I don't know of other cases that use our style of public-interface and private-impl (I mean, it's classic, I just don't think I've seen it outside our code base in a Dart package).

The reason for that could be that Dart doesn't *need* it. In Java, you'll want to define an interface, and an implementation of that interface, because otherwise you don't have an interface at all.
In Dart, with open classes, you can just write the implementation, and then derive an interface from it anyway.

The only real reason to have a public interface and private implementation is to have public members in the implementation, which are not part of the interface, and consider them "pseudo hidden" that way. It's just rarely done, because you can just make private members instead, and not worry about them getting exposed at all.

**Comment by bwilkerson** [comment]
> The only real reason to have a public interface and private implementation is to have public members in the implementation, which are not part of the interface, and consider them "pseudo hidden" that way. It's just rarely done, because you can just make private members instead, and not worry about them getting exposed at all.

Yep. But in our case we want those "pseudo hidden" APIs to be usable outside the library, but within our package.

**Comment by scheglov** [comment]
If we merge interface and implementation classes, we would start using `sealed`.
https://dart-review.googlesource.com/c/sdk/+/308402

**Comment by scheglov** [comment]
Hm... Looking at it some more.
```dart
sealed class ExtendsClause implements AstNode { ... }

final class ExtendsClauseImpl extends AstNodeImpl implements ExtendsClause { ... }
```

`ExtendsClause` is `sealed`, so we can prove that any `ExtendsClause` is also `ExtendsClauseImpl`.

Currently, in the analyzer resolution visitors we have to rely on `covariant`, and write `visitExtendsClause(covariant ExtendsClauseImpl node)`. I wonder if it would have been good to allow to write just `visitExtendsClause(ExtendsClauseImpl node)`, i.e. without `covariant` to say: I want to handle this node as ExtendsClauseImpl here because I'm going to write some of its properties. Well, maybe this is too analyzer specific, and does not have wide use outside.

**Comment by jakemac53** [comment]
> In that sense, I'm arguing that someone writing an exhaustive switch voluntarily made a choice similar to that of implementing a class. They are aware that their code will need updating when a new type is added. And if they don't want this, they can opt out by making a nonexhaustive switch (using `_`).

This issue with this approach, is that all your users get broken when a new type is added and they `pub upgrade`. To avoid that you would have to maintain a tight version constraint on `analyzer`, and have a contract around that (lets say new subtypes are only added in minor versions).

But we don't have any good linting to ensure people use this type of constraint today, or even good patterns established and agreed upon.

**Comment by jakemac53** [comment]
Fwiw, I would love it if pub had a diagnostic that would warn if you don't constrain yourself to minor release ranges, whenever you do a variety of things in this category, such as:

- switch on a sealed class you don't own
- implement a class you don't own
- override a method from a class you don't own

etc

**Comment by rrousselGit** [comment]
+1 on the lint.


I've used tigher version constraints before.  I voluntarily "implement" AstVisitor on purpose in `custom_lint`. Because I **do** want a compilation error if a method is not implemented. And I use tighter constraints for this.





I think a lint rule and considering adding new sealed subtypes as minor bumps would be ideal. 


I find it very counter productive if package authors voluntarily avoid using `sealed` for the sake of versioning.

**Comment by scheglov** [comment]
Why a minor version bump?
If a change is de-facto breaking, it should be a major version bump.
The clients will adapt to the new major version, one way or another.

**Comment by lrhn** [comment]
Whether a change is breaking, in the semver perspective, it's not necessarily as simple as "can any code possibly stop working?"

It's more like "will all intended uses keep working?", where the critical point is how well you've signaled which uses are intended.

Breaking an unintended, and therefore unsupported, use is not your problem.
If you never promised that it should work, not working is within specification.

Changing the API of a class that was never intended to be implemented or extended may break someone doing it anyway, but it's not semver-breaking if it doesn't break the supported API usage that's actually being versioned. The publisher decides what that is. (The rest is a documentation problem.)

In practice it's rarely as clear cut, not if your part of a large organization that tells you who you can break out not, but it is a perspective that a single, third-party developer can choose to take with their package.

That is, "de facto breaking" is only something you should care about if you've promised not to cause any possible breaking, and at that point it's actually "de jura" breaking too.

**Comment by scheglov** [comment]
When we make a class `sealed`, we promise: here are all its subtypes, you can use exhaustive switches.
Adding a new subtype will result in compile-time errors for these switches.

**Comment by rrousselGit** [comment]
> Or adding an optional named parameter to a method (breaking if somebody is implementing or extending and overriding that method).

Heck, even for global functions, that's breaking too if you consider things like type inference or `if (fn.runtimeType == void Function()` or toString.  
There are way too many ways something could "break". 😝 

_insert mandatory [xkcd reference](https://xkcd.com/1172/)_ 
![image](https://github.com/dart-lang/language/assets/20165741/68f09a0d-7e5b-4f08-985c-670f800b47ac)


____



In the end, I think it's about what makes the most sense for users.

If semantic versioning is going to prevent packages from using `sealed` because that's too inconvenient to make  "breaking changes", then I'd rather stop using semver.  
Flutter and Dart don't use semantic versioning after all. 


Ultimately, with exhaustive pattern matching, the key point for me is that it's a conscious decision made by the package users. 

Folks have the choice to write either: 

```dart
switch (Base()) {
  case Impl1():  print('');
  case Impl2(): print('');
}
```

Or:

```dart
switch (Base()) {
  case Impl1():  print('');
  case _: print('');
}
```


Folks who opted for the exhaustive check did so on purpose. They **want** the "breaking change".   
Folks who don't like the idea can use a non-exhaustive switch.  
Users have the choice and made their decision.   

If packages refuse to offer exhaustive checks for the sake of versioning,  what's the point of having the `sealed` keyword in Dart?

**Comment by rrousselGit** [comment]
Just passing by to say thanks for making various Ast classes "sealed" ❤️
